src/parser/ParsedQuery.cpp

```diff
@@ -130,6 +120,22 @@ Variable ParsedQuery::addInternalBind(
 
 // ________________________________________________________________________
 void ParsedQuery::addSolutionModifiers(SolutionModifiers modifiers) {
+  auto checkUsedVariablesAreVisible =
+      [this](const sparqlExpression::SparqlExpressionPimpl& expression,
+             const std::string& locationDescription) {
+        for (const auto* var : expression.containedVariables()) {
+          // TODO: think of a solution to make this work with ConstructClause as
+          //  well.
```

Comment:
Well, one solution would be to store the visible variables in the `ParsedQuery` and not in the `SelectClause`. But then we cannot use the `SelectClause` as a standalone without the parsed query, and that would be a shame. What about letting the `SelectClause` and the `ConstructClause` both store the visible variables? With inheritance etc. we don't even need code duplication:

```
struct ClauseBase {
  std::vector visibleVariables_;
  // possibly getter and setter functions.
};

class SelectClause : ClauseBase {
}; 
class ConstructClause : ClauseBase {
};
std::variant<Selectclause, Constructclause> ...
// In a std::visit you can now use the inherited functionality from the class base for both of them.
```